doc/pandas-2.0/source/internal-architecture.rst

```diff
@@ -0,0 +1,714 @@
+
+If we add more composite data structures (Categorical can be thought of as
+one existing composite data structure) to pandas or alternate non-NumPy data
+structures, there will be cases where the semantic information in a Series
+cannot be adequately represented in a NumPy array.
+
+As one example, if we add pandas-only missing data support to integer and
+boolean data (a long requested feature), calling ``np.asarray`` on such data
+may not have well-defined behavior. As present, pandas is implicitly converting
+these types to ``float64`` (see more below), which isn't too great. A decision
+does not need to be made now, but the benefits of solving this long-standing
+issue may merit breaking ``asarray`` as long as we provide an explicit way to
+obtain the original casted ``float64`` NumPy array (with ``NaN`` for NULL/NA
+values)
+
+For pandas data that does not step outside NumPy's semantic realm, we can
+continue to provide zero-copy views in many cases.
+
+Missing data consistency
+========================
+
+Once the physical memory representation has been effectively decoupled from the
+user API, we can consider various approaches to implementing missing data in a
+consistent way for every logical pandas data type.
+
+To motivate this, let's look at some integer data:
+
+.. ipython:: python
+
+   s = pd.Series([1, 2, 3, 4, 5])
+   s
+   s.dtype
+   s.values
+
+If we assign a ``numpy.NaN``, see what happens:
+
+.. ipython:: python
+
+   s[2] = np.NaN
+   s
+   s.dtype
+   s.values
+
+The story for boolean data is similar:
+
+.. ipython:: python
+
+   s = pd.Series([True, False, True])
+   s.dtype
+   s[2] = np.NaN
+   s.dtype
+   s.values
+
+This implicit behavior appears in many scenarios, such as:
+
+* Loading data from any source: databases, CSV files, R data files, etc.
+* Joins or reindexing operations introducing missing data
+* Pivot / reshape operations
+* Time series resampling
+* Certain types of GroupBy operations
+
+A proposed solution
+~~~~~~~~~~~~~~~~~~~
+
+My proposal for introducing missing data into any NumPy type outside of
+floating point (which uses ``NaN`` for now) and Python object (which uses
+``None`` or ``NaN`` interchangeably) is to **allocate and manage an internal
+bitmap** (which the user never sees). This has numerous benefits:
+
+* 1 byte of memory overhead for each 8 values
+* Bitmaps can propagate their nulls in C through bitwise ``&`` or ``|``
+  operations, which are inexpensive.
+* Getting and setting bits on modern hardware is CPU-inexpensive. For
+  single-pass array operations (like groupbys) on large arrays this may also
+  result in better CPU cache utilization (fewer main-memory reads of the
+  bitmap).
+* Hardware and SIMD "popcount" intrinsics (which can operate on 64-128 bits at
+  a time) can be used to count bits and skip null-handling on segments of data
+  containing no nulls.
+
+Notably, this is the way that PostgreSQL handles null values. For example, we
+might have:
+
+.. code-block::
```

Comment:
This code block is not showing up in the generated HTML document. Is a language required?